SupportsFeatureMixin

durandom opened the pull request:
This is a first stab at refactoring `is_available?` and `validate_` calls.

Before this a `validate_<feature>` call had to be implemented as an instance method and had to return a hash or nil. `{ available: true, message: 'Not available' }`
In case nil is returned, the feature is not available. 
In case a Hash is returned and available is false, then the feature is basically available, but for the reason in :message its currently not. 
If :available is true, the feature is available right now.
My concern with this is 
* wording: validate_ sounds like rails validation, but we are not validating the object itself
* return value: either a nil value could be returned or a hash
* an object/record needs to be present to check the general validity of a feature

The `validate_` method would generally be used by the UI to test the availability of an operation/feature. 
This would be done by a pair of methods: `is_available?(operation)` and `is_available_now_error_message(operation)`
My concern with this is:
* wording: `is_available?` does not communicate what is available and the side effect that it depends on internal state of the object
* input is keyword which cannot be introspected like a method without any arguments (what can I query for?)
* an object/record needs to be present to check the general validity of a feature

A first refactor was the extraction into a [AvailabilityMixin](https://github.com/ManageIQ/manageiq/pull/7062) and a rename of those `available` method names: https://github.com/ManageIQ/manageiq/pull/3596

I propose a more DSL like approach that adds a `supports` class method, which defines 3 methods
* A class level method that returns the general support of a feature
* An instance level method that answers if the feature is supported by this instance
* Another instance level method that returns the reason for being not supported by the instance

```ruby
class Animal
  supports :die
  supports :fly, false
end

class Bird < Animal
  supports :fly do |bird|
    if bird.penguin? 
      "Penguins can't fly"
    else
      true
    end
  end
end

Animal.supports_die? # true
Animal.new.supports_die? # true
Animal.supports_fly? # false
Bird.supports_fly? # true
Bird.new(penguin: true).support_fly? # false
Bird.new(penguin: true).unsupported_fly_reason # 'Penguins can't fly
```

As an example I refactored `live_migrate`

@Fryguy @martinpovolny @mfalesni what do you think of this?

@miq-bot add_label refactoring

durandom commented:
/cc @blomquisg

mfalesni commented:
@durandom Hey, you probably meant @mzazrivec ? :)

durandom commented:
@mfalesni yes, sorry, so: @mzazrivec what do you think of this approach?

chessbyte commented:
@durandom LOVE ❤️ the idea!  

Would be nice to memoize the unsupported reason for a particular operation.

/cc @abonas @simon3z

Fryguy commented:
I also love the general idea.  However, there's something about the DSL itself that feels wrong, but I can't quite put my finger on it.  I think it's the fact that returning a String from the block, which is usually truthy, is interpreted as falsey, and you have to return a literal true.  Additionally, the `supports :die, false` feels weird as well, as that false gets visually lost when you pass a block.  I wonder if it's better to just have another method like `unsupported :die` (I was trying to come up with the "opposite" of `supports`, and couldn't figure out the right word).

Even so, I could see implementing this as is, and we can change it later. Let's discuss...I'm very excited by this :)

Fryguy commented:
Perhaps not implementing `supports` for a particular term automatically assumes it's false?  Then there's no ambiguity.  The downside is that a typo will pass through, but hopefully the caller would catch that with proper tests.

```ruby
class Animal
  supports :die
end

Animal.supports_fly? # false
Animal.supports_diet? # false (example of typo case)
```

chessbyte commented:
@Fryguy
```ruby
class Animal
  supports :die
  unsupports :fly  # non-English, unfortunately
  does_not_support :fly  # English, but wordy
  unimplemented :fly # English, but no relation to supports
end
```

chessbyte commented:
@Fryguy with your approach, you lose the ability to override something supported on a base class and not implemented on a subclass.

Fryguy commented:
@chessbyte Good point.

durandom commented:
thanks for the valuable and great input everybody

> Would be nice to memoize the unsupported reason for a particular operation.

I thought about that too, but then memoizing is probably only really valuable if you have an expensive operation that checks whether the feature is supported or not. And then it might have changed in the meantime (like query an external service). So you might get a false positive. But then, this is probably an edge case too.

If you think it's worth of transparently memoizing the result, I'll add it. I am a bit undecided and would not add it until the first expensive check comes along.

durandom commented:
> However, there's something about the DSL itself that feels wrong. I think it's the fact that returning a String from the block, which is usually truthy, is interpreted as falsey, and you have to return a literal true. 

Yes, that was annoying me too. But we want to _conditionally_ support something. So
```ruby
supports :fly # generally supported
# somewhere down or in a subclass
unsupported :fly do |animal|
  if animal.wings_present?
    "no wings, no fly"
  end
end
```

this leaves the decision at multiple places (it could be even more than 2) which makes it harder to debug. And what comes first? Some [Apache mod_access order confusion?](https://httpd.apache.org/docs/current/mod/mod_access_compat.html#order)
So I would strongly opt for having it in one place, which leaves us with a block that has to return "Yes, I support it" or "Nope, I can't because..."

1. return a special object? I think then we are back to the return Hash style
2. return true/reason string. As its now
3. return nothing, but set an internal object (like ActiveModel::Errors)

Actually the 3rd approach could lead to some interesting added value. Like memoizing :point_up: and querying for a bunch of features in on go.

durandom commented:
> Additionally, the supports :die, false feels weird as well, as that false gets visually lost when you pass a block.

How about having two DSL methods: `supported :feature, &block` and `unsupported :feature`. 
The latter one intentionally disallows a block (because conditionally unsupported screams for double negation, which screams for bugs)

I'd also raise an exception in case a feature has more than 2 support definition (to avoid the what comes first problem)

chrisarcand commented:
@durandom Do we need to meta define methods instead of a simple hash lookup? `Animal.supports?(:fly)` ?

This seems more flexible (can open to multiple params for multiple checks, whatever) and overall much less needless meta (easier to grep, etc).

durandom commented:
@chrisarcand I actually prefer methods without arguments over methods with arguments :smile: But your argument towards multiple checks at once is valid. Although probably rarely used? (not sure)

Then I also like, that a type raises a `NoMethod` error, whereas an undefined argument raises a special error / KeyError?

But yeah, I see that approach has positive effects though.
What do you think of using the ActiveModel::Errors API for that?

```ruby
vm.unsupported = ActiveModel::Errors.new
vm.unsupported.include?(:fly)
vm.supports?(:fly) # false, because it's in `unsupported`
```

durandom commented:
> Do we need to meta define methods instead of a simple hash lookup?

Oh, and methods give us inheritance / overriding methods for free, as @chessbyte noted. With a hash lookup this would require more setup, I guess

chrisarcand commented:
> Then I also like, that a typo raises a NoMethod error, whereas an undefined argument raises a special error / KeyError?

I don't like either ^_^ Since this is a custom DSL, I'd rather see a custom `NotSupported` error.

> What do you think of using the ActiveModel::Errors API for that?

I guess more hashing out would need to be presented first, but first impression is that it seems unnecessary, tbh. Maybe?

> Oh, and methods give us inheritance / overriding methods for free

Wouldn't overriding `::supports` on the subclass do the same thing? Or are you saying you want to override things _per instance_ (😬 , that seems beyond the scope of what I thought the point of this was)

durandom commented:
>  Or are you saying you want to override things per instance

Yes, per instance. Maybe the Animal/Bird example was bad to show that.
```ruby
Vm.where(disconnected_from_ems: true).supports?(:migrate) # => false
Vm.where(disconnected_from_ems: false).supports?(:migrate) # => true
```

martinpovolny commented:
@durandom : I'd say that the approach is good. Seems to be sufficient for the task at hand. :+1:

mzazrivec commented:
@durandom I'm OK with the general direction. Nice work!

chrisarcand commented:
> Yes, per instance.

Ahh I see. Well I guess if we have a need for the flexibility I mentioned this shouldn't be ridiculous to track down and change in the future anyway. 👍 

My last little comment is that I think the API of 'string for reason why it's unsupported, true otherwise' is a little weird looking at it by itself on classes, (like, I would be a little confused understanding the purpose just looking at that block) but I don't have a better alternative off the top of my head 🤐

I'm spread a little thin and won't comment further. This is a great direction for sure, nice work @durandom!  👍 👍 ❤️

blomquisg commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -0,0 +1,56 @@
+          true
+        else
+          false
+        end
```
Feels like this class method definition could be reduced to:

```ruby
self.class.send(:define_method, method_name) do
  is_supported
end
```

* `result` is either `true`, or `reason`
* `reason` is never `=== true` (unless someone passes `reason = true` in the arguments, which seems weird)
* the last if statement can be reduced to `return result === true`
* `result === true` only if `is_supported == true`
* so, `return result === true` is the same as `return is_supported`

Is that right?

chrisarcand replied:
Yup.

This stems from my comment [here](https://github.com/ManageIQ/manageiq/pull/8886#issuecomment-221301325). It's intended to strong compare `=== true` as opposed to 'truthy' (weird). I suggest we separate the concern of whether or not something is supported (bool) and the reason why it isn't. This involves returning either a custom result object (make separate class or Struct of some variety) or at minimum a multivalue array (`[true, nil]`, `[false, "Reason why it's not supported"]`)

chrisarcand replied:
(For fun, but I still think we should change the API to return something more informative)

```
self.class.send(:define_method, method_name) do
  result = if is_supported
             true
           else
             reason
           end
  if result === true
    true
  else
    false
  end
end

# =>

self.class.send(:define_method, method_name) do
  result = is_supported || reason   # Changed
  if result === true
    true
  else
    false
  end
end

# =>

self.class.send(:define_method, method_name) do
  result = is_supported || reason
  result === true  # Changed
end

# =>

self.class.send(:define_method, method_name) do
  is_supported   # Changed. Reason removed because it’s never === true
end
```

blomquisg commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -0,0 +1,56 @@
+        else
+          false
+        end
+      end
```
It looks like this could be:

```ruby
define_method(method_name) do
  block ? block.call(self) : is_supported
end
```

blomquisg commented:
@durandom do you want to pull in some more VM features into this?  Or, move forward with the `control` feature and add more features in follow up PRs?

I'm fine either way.  Just want to make sure I'm not jumping the gun...

durandom commented:
@blomquisg you are right, that those methods can be reduced quite a bit in size.

And I think I try to rework this a little bit to get rid of this awkward "return true or string" in the DSL.

durandom commented on spec/models/mixins/supports_feature_mixin_spec.rb:
```diff
@@ -0,0 +1,96 @@
+  end
+
+  it "propagates deeply nested supports" do
+    expect(Post.supports_nested?).to be true
```
this one is failing, because I dont propagate up deeply nested includes.

@blomquisg @Fryguy @chessbyte do we want to keep the current codebase layout of nested modules?
If I'd put all `supports` methods into either the class `Vm` or one shared module `VmOperations` this would work.

durandom replied:
ok, never mind. I switched to another, much simpler approach. stay tuned :smile:

durandom commented:
@miq-bot add_label wip

What do you think of this API?
https://github.com/durandom/manageiq/blob/c5f2e46aaa9a46e7f5277a58d2ce36d883f794e7/spec/models/mixins/supports_feature_mixin_spec.rb#L2-L43
Also see the examples in the spec how to use it.

I removed the awkward "return true or string" block and replaced it with something that is inspired by ActiveModel::Errors 
Right now it's just a Hash, but it could be a custom class in case we need more functionality

durandom commented:
I reworked it once again to handle the nested modules better.
See https://github.com/durandom/manageiq/blob/daffc851734dc674adff9a597a02d9a374d99a83/spec/models/mixins/supports_feature_mixin_spec.rb

We wrap the `supports` calls in an `included` block. This way they get added to the Model and not to the Module first.
Now it's required to extend every module in the "include tree" with `ActiveSupport::Concern` (this handles handing down the `included` blocks. 

Aside from naming things `supports`, `supports_not` and `.unsupported[]` I think this is ok.
I'm curious on your feedback

@miq-bot remove_label wip

Fryguy commented:
@durandom I haven't reviewed yet, but quickly saw the spec, and just wanted to point out that the spec should nit be defining classes, otherwise it lends itself to cross test contamination.  Instead, if you need a class, use Class.new in a before block.  There are examples of this in other specs

durandom commented:
@Fryguy thanks for the note about the Class.new

I also cleaned up the spec to be much more precise and informative :smile: 
Also added doc to SupportsFeatureMixin and squashed into two commits.

I'm still open to wording the DSL. This will be just a matter of search and replace

chessbyte commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -0,0 +1,96 @@
+  #
+  #   Post.unsupported[:publish]                   # => nil
+  #   Post.unsupported[:fake]                      # => "We keep it real"
+  #   Post.new(featured).unsupported[:archive]     # => "Its too good"
```
Somehow this Hash interface for `unsupported` feels odd.

Why should the caller know it is a hash?  Caller probably wants to call it as a method: `Post.unsupported(:publish)`.  

Additionally, maybe slightly changing the name of this method to `unsupported_reason` may make it clearer.

durandom replied:
+1 makes sense.
I think I'll move setting the reason to a private method then and retrieving it to a public

durandom commented:
I moved `unsupported` to `unsupported_reason(feature)` and added `unsupported_reason_add(feature, reason)`
The latter is a private method. Therefore the block in `self.supports` is evaluated in instance context.

see the last commit.

blomquisg commented:
@martinpovolny, @Fryguy, and @durandom, let's see what we can agree on with this PR.  I'd like to come to an agreement on the API and the semantics.  There's no immediate rush, but I do want to move this one forward.

Also, lemme know if you have anyone else that should look this over.

martinpovolny commented on app/models/vm_or_template/operations.rb:
```diff
@@ -93,6 +95,25 @@ def validate_vm_control
+  included do
+    supports :control do
+      msg = if vm.retired?
+              'The VM is retired'
```
`i18n` here and below?

martinpovolny replied:
Probably just `N_()` and have the caller call `_()` if needed. But seems we need these strings in the catalog.

mzazrivec replied:
I think we need to use `_()` here.

martinpovolny replied:
@mzazrivec : I was thinking that we might need to just mark this here and do the translation in the caller unless the caller is the API. But I did not look into the details.

martinpovolny commented on app/models/vm_or_template/operations/relocation.rb:
```diff
@@ -1,4 +1,10 @@
+  extend ActiveSupport::Concern
+
+  included do
+    supports_not :live_migrate, "Live Migrate VM Operation is not available for VM or Template."
```
same here -- `N_()`

mzazrivec replied:
We need to use `_()`, not `N_()`

martinpovolny commented:
@durandom, @blomquisg : I like the pattern. It looks good. Let's start using it and see if it lasts.

durandom commented:
I moved the example implementation for live_migrate to another PR

@Fryguy @chessbyte how do you like this? If you are ok, merge?

durandom commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -0,0 +1,113 @@
+
+    def unsupported
+      # TODO: durandom - is Thread.current needed here?
+      Thread.current["unsupported_#{object_id}"] ||= {}
```
@jrafanie I'm not so sure about the current state of threading. This has to be a variable on the class level. My ancient knowledge is, that it's not thread safe and therefore we should use `Thread.current`. Is that still valid?

jrafanie replied:
Do we want this to be threadsafe?  Does it make sense to for a feature to be supported in one thread and not another?

Maybe https://github.com/ruby-concurrency/concurrent-ruby threadsafe hash or map could be used?

Maybe we punt and use this for now?  I don't know.

durandom replied:
> Do we want this to be threadsafe? Does it make sense to for a feature to be supported in one thread and not another?

No, its the same in every thread, but that Hash gets populated upon calling the class method. Which would be ok, if we'd eager_load all classes. But we don't. So it could be that thread1 loads the class and modifies the hash, while another one does too.

Good pointer to concurrent-ruby. As we use it already, I'll go with the threadsafe hash, thanks @jrafanie

durandom commented:
@Fryguy @chessbyte and whoever it might concern: Before we roll this out all over our code, do you think this API is good?

blomquisg commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -0,0 +1,114 @@
+  #
+  #   Post.unsupported_reason(:publish)                   # => nil
+  #   Post.unsupported_reason(:fake)                      # => "We keep it real"
+  #   Post.new(featured).unsupported_reason(:archive)     # => "Its too good"
```
`featured: true`, not just `featured`, I think.  So, this example should read:

```ruby
#   Post.new(featured: true).unsupported_reason(:archive)     # => "Its too good"
```

chessbyte commented:
I am 👍 on this PR.  The DSL syntax looks good to me.

blomquisg commented:
I'm good with this API.  It's a good replacement for `is_available?` I think:

```ruby
# old
@record.is_available?(:feature)

# new
@record.supports_feature?
```

durandom commented:
cool. If @Fryguy gives his :+1: we could merge :smile:

Fryguy commented:
I'm not a fan of supports_not as a phrase, but overall this is a great change, so I'm going to merge.  If we decide later the wording is bad, it's not hard to find/replace something like that.

chrisarcand commented:
Nice work @durandom

durandom commented:
@miq-bot add_label darga/yes

@chessbyte so we backport this as per https://github.com/ManageIQ/manageiq/pull/9272#issuecomment-231861754

I think there are more PRs coming, that will depend on this. And it's not really a change in logic, but an addition of a feature

durandom commented:
@miq-bot remove_label darga/no

chessbyte commented:
Darga backport details:
```sh
commit d925ef0a39bd826a497bcc95c0433fec402dc7b4
Author: Jason Frey <fryguy9@gmail.com>
Date:   Tue Jun 21 17:31:31 2016 -0400

    Merge pull request #8886 from durandom/supported_concern
    
    SupportsFeatureMixin
    (cherry picked from commit 80e2c93c270909aec168ae9293a10c0d37ceb2d0)
```
